### PR TITLE
feat: add flox-nixpkgs fetcher

### PR DESCRIPTION
**Pull request by aakropotkin**
## Proposed Changes

- Adds a new type of fetcher, `flox-nixpkgs:v0/<REF-OR-REV>`, which _wraps_
  upstream `nixpkgs` with additional configuration.
  + Currently injects `config.allowBroken = true`, and
    `config.allowUnfree = true` when evaluating `nixpkgs`.
- Fetcher scheme includes a versioning mechanism to allow us to
  safely roll out breaking changes to the fetcher over time.
  + Starts with `v0`.
- Fetchers fingerprints do NOT conflict with _unwrapped_ `github`
  fetchers for `nixpkgs` at the same revisions.
- This fetcher's _locked references_ are safe to include lockfiles.
- This fetcher's _installable URIs_ are suitable for use with `buildenv`.
- fix: `iwyu` wrapper for `pkgdb`.

## Release Notes

N/A

### CODE REVIEW

**Review by ysndr — COMMENTED** [review]
Haven't yet looked more closely into this, but hte first thing i noticed, is that we're still _replacing_ "github" with `flox-nixpkgs`, rather than _wrapping_ a generic flakeref.

I.e.:

```
flox-nixpkgs+<flakeref_url>[&|?]<additional attrs>


github:nixos/nixpkgs -> flox-nixpkgs+github:nixos/nixpkgs?filters=...
```

I don't know exactly how internal these urls are (they do end up in the lockfile iirc), but extending this to also support other origins than upstream nix will be far easier if a flakeref was generically wrapped.

**Comment by ysndr** [comment]
There might be implementation details I'm missing, but shouldnt we be able to parser a `flox-nixpkgs+<flakeref>` using a custom fetcher that
1) strips `flox-nixpkgs+`
2) strips `flox-nixpkgs related attributes`
3) fetches <flakeref>
4) assumes <flakeref> is a `nixpkgs` flake applies flox customization


> Whats the filter parameter you've got in your example?

just an example for flox specific attributes (i guess it was `rules` or so before (i see this pr does not include any of the rules stuff, but it was in the original one)

**Comment by aakropotkin** [comment]
That is an alternative approach that would be more general purpose. It's not what I chose to do here though.

My major concern was ensuring modifications could be reproduced over time, so something more general purpose seemed more challenging with that goal in mind.

File: pkgdb/tests/flox-nixpkgs.bats [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  assert_output "$URL";
+
+}
+
+
+# ---------------------------------------------------------------------------- #
+#
+#
+#
+# ============================================================================ #
```
Comment (mkenigs) [inline pkgdb/tests/flox-nixpkgs.bats:82]: suggestion: add a test that an `outPath` from the wrapped flake is the same as one from the plain flake

Comment (mkenigs) [reply pkgdb/tests/flox-nixpkgs.bats:99]: Thanks!

File: pkgdb/src/fetchers/wrapped-nixpkgs-input.cc [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+
+  /* Lock the filled out template to avoid spurious re-locking and silence the
+   * "Added input ..." message. */
+  flox::NixState           nixState;
+  nix::ref<nix::EvalState> state = nixState.getState();
+  nix::FlakeRef wrappedRef = nix::parseFlakeRef( "path:" + tmpDir.string() );
+  /* Push verbosity level to suppress "warning: creating lock file ..." */
+  auto oldVerbosity = nix::verbosity;
+  nix::verbosity    = nix::lvlError;
+  auto _locked      = nix::flake::lockFlake( *state, wrappedRef, {} );
```
Comment (billlevine) [inline pkgdb/src/fetchers/wrapped-nixpkgs-input.cc:101]: Should we be using the new wrapper `flox::lockFlake` to ensure downloads (if any) occur in a child process?

**Review by billlevine — CHANGES_REQUESTED** [review]
Just the one comment regarding the `lockFlake`.  Could cause intermittent crashes if a download happens as a result.

File: pkgdb/src/fetchers/wrapped-nixpkgs-input.cc [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+  input.attrs.insert_or_assign( "type", this->type() );
+
+  auto path = nix::tokenizeString<std::vector<std::string>>( url.path, "/" );
+
+  if ( path.size() != 2 )
+    {
+      throw nix::BadURL( "URL '%s' is invalid", url.url );
+    }
+
+  if ( path.size() != 2 )
```
Comment (zmitchell) [inline pkgdb/src/fetchers/wrapped-nixpkgs-input.cc:303]: This is duplicated

File: pkgdb/src/fetchers/wrapped-nixpkgs-input.cc (hunk 2 above)
Comment (zmitchell) [reply pkgdb/src/fetchers/wrapped-nixpkgs-input.cc:101]: Yes

**Comment by zmitchell** [comment]
@mkenigs shepherd away

**Comment by mkenigs** [comment]
> @mkenigs shepherd away

Haha okay I'll make the requested changes and then re-request your review

File: pkgdb/src/fetchers/flake-v0.nix.in [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+# ---------------------------------------------------------------------------- #
+
+    lib = {
+
+      # eachDefaultSystemMap fn
+      # -----------------------
+      # Given a function `fn' which takes system names as an argument, produce
+      # an attribute set whose keys are system names, and values are the result
+      # of applying that system name to `fn'.
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkgdb/src/fetchers/flake-v0.nix.in]: Fixed

File: pkgdb/src/fetchers/flake-v0.nix.in [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }; # End `config'
+
+
+# ---------------------------------------------------------------------------- #
+
+    lib = {
+
+      # eachDefaultSystemMap fn
+      # -----------------------
+      # Given a function `fn' which takes system names as an argument, produce
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkgdb/src/fetchers/flake-v0.nix.in]: Fixed

File: pkgdb/build-aux/iwyu [hunk 6]
```diff
@@ -162,7 +164,16 @@ done
...
 # Set fallbacks.
 
-: "${_PROJECT_ROOT:=$PWD}"
+if [[ -z "${_PROJECT_ROOT:-}" ]]; then
+  if [[ -d "$PWD/.git" ]]; then
+    _PROJECT_ROOT="$PWD"
+  elif [[ -d "$PWD/../.git" ]]; then
+    _PROJECT_ROOT="${PWD%/*}"
+  else
+    _PROJECT_ROOT="$( $GIT rev-parse --show-toplevel; )"
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply pkgdb/build-aux/iwyu]: Fixed

File: pkgdb/src/fetchers/wrapped-nixpkgs-input.cc (hunk 2 above)
Comment (mkenigs) [reply pkgdb/src/fetchers/wrapped-nixpkgs-input.cc]: Fixed

File: pkgdb/tests/flox-nixpkgs.bats [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  assert_output "true";
+}
+
+
+# ---------------------------------------------------------------------------- #
+
+@test "'flox-nixpkgs' fetcher sets 'allowUnfree' and 'allowBroken'" {
+  run --separate-stderr "$PKGDB_BIN" eval "let
+    nixpkgs = builtins.getFlake
+                \"flox-nixpkgs\:v$FLOX_NIXPKGS_VERSION/$NIXPKGS_REV\";
```
Comment (aakropotkin) [inline pkgdb/tests/flox-nixpkgs.bats:30]:
```suggestion
                \"flox-nixpkgs:v$FLOX_NIXPKGS_VERSION/$NIXPKGS_REV\";
```

**Review by zmitchell — APPROVED** [review]

**Pull request merged**